lib/railties/all/railties.rbi

```diff
@@ -1,18 +1,1274 @@
 # typed: strong
 
 module Rails
-  sig {returns(Rails::Application)}
-  def self.application; end
-  
-  sig { returns(Pathname) }
-  def self.root; end
+  extend(::ActiveSupport::Autoload)
+
+  class << self
+    def app_class; end
+    def app_class=(_arg0); end
+
+    sig { returns(Rails::Application) }
+    def application; end
+
+    def application=(_arg0); end
+    def autoloaders; end
+
+    sig { returns(ActiveSupport::BacktraceCleaner) }
+    def backtrace_cleaner; end
+
+    sig { returns(ActiveSupport::Cache::Store) }
+    def cache; end
+
+    def cache=(_arg0); end
+    def configuration; end
+
+    sig { returns(ActiveSupport::StringInquirer) }
```

Comment:
While technically more accurate, I'd argue this is a bad change because now `Rails.env.development?` and similar methods do not pass typechecking even though they should.

@jeffcarbs do you have a fix for this that you use, or are any files that use `Rails.env.development?` just marked false?

Reply:
@connorshea - we have the following RBI in a “manual” RBI file:
```
class ActiveSupport::StringInquirer
  sig { returns(T::Boolean) }
  def development?; end

  sig { returns(T::Boolean) }
  def production?; end

  sig { returns(T::Boolean) }
  def staging?; end

  sig { returns(T::Boolean) }
  def test?; end
end
```

I do see how this could be annoying though. If you want to remove this sig and maybe leave a comment as to why it’s omitted I won’t be offended 🙂 I can just add it as RBI in my repo.

Reply:
hmm, maybe add those signatures as a comment in the railties.rbi above the current `env` signature? That way we can lead people toward what they _should_ be doing? I definitely like having the methods typecheck correctly and your solution seems pretty good.

Reply:
I’ve done that in other RBI where you can’t really make assumptions about other people’s setup. For instance, sidekiq defines rails extensions but they’re only relevant if you’re using rails. If you typed them in the sidekiq.rbi it would incorrectly tell sorbet ActiveRecord was there. For example: https://github.com/sorbet/sorbet-typed/blob/master/lib/sidekiq/all/sidekiq.rbi#L102-L105

Reply:
This returns an `ActiveSupport::EnvironmentInquirer` from Rails 6.1. See https://github.com/rails/rails/pull/38657

We should return the correct type in those cases, but we should also do this:

```ruby
class ActiveSupport::EnvironmentInquirer
  sig { returns(T::Boolean) }
  def development?; end

  sig { returns(T::Boolean) }
  def production?; end

  sig { returns(T::Boolean) }
  def test?; end
end
```

Reply:
in my opinion it's also okay to ship this to everyone:

```ruby
class ActiveSupport::StringInquirer
  sig { returns(T::Boolean) }
  def development?; end

  sig { returns(T::Boolean) }
  def production?; end

  sig { returns(T::Boolean) }
  def staging?; end

  sig { returns(T::Boolean) }
  def test?; end
end
```

technically, *any* method that ends with `?` when called on an `ActiveSupport::StringInquirer` will return a `T::Boolean`. unlike the sidekiq example this doesn't introduce new incorrect dependencies.

Reply:
Ah you're right, it will technically never raise
```ruby
> Rails.env.womp?
=> false
```

I think this is one of those places where the RBI can enforce things that aren't enforced at runtime. For example, while the string/environment inquirer technically won't raise if you ask about a bad env, it would be nice for sorbet to catch the typo. Otherwise you might think you're checking for a particular environment when you actually aren't 😬 

How about this: since `production` `test` and `development` are default environments (i.e. created when you `rails new`), we can ship RBI for them. So:

```
class ActiveSupport::EnvironmentInquirer # And StringInquirer depending on Rails version
  sig { returns(T::Boolean) }
  def development?; end

  sig { returns(T::Boolean) }
  def production?; end

  sig { returns(T::Boolean) }
  def test?; end
end
```

I think we can leave a comment for people to add other environments (e.g. `staging`) if they need them. We actually have a few other environments so would need to do that anyway.